### PR TITLE
chore: define vertex spec for user define source

### PR DESCRIPTION
**Pull request by KeranYang**
Also re-arrange our e2e test suites to avoid CI workflow time out.

### CODE REVIEW

File: pkg/reconciler/pipeline/validate.go [hunk 1]
```diff
@@ -89,6 +89,17 @@ func ValidatePipeline(pl *dfv1.Pipeline) error {
 		return fmt.Errorf("pipeline has no sink, at least one vertex with 'sink' defined is required")
 	}
 
+	for k, s := range sources {
+		if s.IsUDSource() {
+			if s.Source.UDSource.Container == nil || s.Source.UDSource.Container.Image == "" {
+				return fmt.Errorf("invalid user-defined source vertex %q, a customized image is required", k)
+			}
+			if s.Source.HTTP != nil || s.Source.Kafka != nil || s.Source.Nats != nil || s.Source.RedisStreams != nil || s.Source.Generator != nil {
```
Comment (vigith) [inline pkg/reconciler/pipeline/validate.go:97]: can you add a TODO to move `redis`, `nats`, to user-defined-source, and not in the core?

**Review by vigith — APPROVED** [review]

File: pkg/apis/numaflow/v1alpha1/source.go [hunk 2]
```diff
@@ -108,3 +113,39 @@ func (s Source) getUDTransformerContainer(mainContainerReq getContainerReq) core
 	}
 	return container
 }
+
+func (s Source) getUDSourceContainer(mainContainerReq getContainerReq) corev1.Container {
+	c := containerBuilder{}.
+		name(CtrUdsource).
+		imagePullPolicy(mainContainerReq.imagePullPolicy). // Use the same image pull policy as the main container
+		appendVolumeMounts(mainContainerReq.volumeMounts...)
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/source.go:121]: Should revisit this.

File: pkg/apis/numaflow/v1alpha1/source.go [hunk 3]
```diff
@@ -108,3 +113,39 @@ func (s Source) getUDTransformerContainer(mainContainerReq getContainerReq) core
...
+		c = c.appendEnv(x.Env...).appendVolumeMounts(x.VolumeMounts...).resources(x.Resources).securityContext(x.SecurityContext).appendEnvFrom(x.EnvFrom...)
+		if x.ImagePullPolicy != nil {
+			c = c.imagePullPolicy(*x.ImagePullPolicy)
+		}
+	}
+	container := c.build()
+	container.LivenessProbe = &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path:   "/sidecar-livez",
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/source.go:141]: Do we have the logic change in `/sidecar-livez`?

File: pkg/reconciler/pipeline/validate.go (hunk 1 above)
Comment (KeranYang) [reply pkg/reconciler/pipeline/validate.go:97]: Could you elaborate more on why?

Comment (vigith) [reply pkg/reconciler/pipeline/validate.go:97]: i want to move those away from our core numaflow code; our core code is getting bloated. i would like to do that with UDSource release (0.11).

File: pkg/apis/numaflow/v1alpha1/source.go (hunk 3 above)
Comment (KeranYang) [reply pkg/apis/numaflow/v1alpha1/source.go:141]: synced offline, I will address this comment in follow-up PRs.

**Review by whynowy — APPROVED** [review]

**Pull request merged**